### DISCUSSION TITLE
[item] all annotation and note content by all tags (combi #241 + #251)

### DISCUSSION THREAD

**Discussion by sjgknight** [windingwind/zotero-better-notes#256] [Note Templates]
## Template Type

Item

## Desired Template Performance

A template where:

    User selects parent items or/and notes
    Template collates all tags from those items 
    Template main note is populated by ##Tag ## Parent item title; ## Note embed (including images and full quote text)

This builds on #251 (which embeds content of annotations/notes, not just links), and #241 (which creates a main note organised by all tags on the notes). 

The below template gets part of the way there, but there are some fixes (which I think are beyond my - try and and see -  capabilities). 

1. The headings of the notes are all H1, which means they aren't effectively navigable by tag.  Ideally they'd either be demoted, or stripped off. 
2. ~~I don't understand how copyNoteImage works (so I get the placeholders/links to the images but they aren't being copied in)~~
3. I'm getting each note twice (which I guess is because both the note and annotation exist, but my memory is #251 doesn't do that). 

The rationale for wanting to do this is that this kind of template helps when we're doing thematic sorts of analysis e.g. to begin the scaffold for the heading structure for literature review (or, in our case, to extract some key themes from the literature). 

```
// @beforeloop-begin
${(() => {
  sharedObj.tagsDict = {};
  sharedObj.updateData = (item, tag) => {
    if (!(tag in sharedObj.tagsDict)) {
      sharedObj.tagsDict[tag] = [];
    }
    sharedObj.tagsDict[tag].push(item);
  };
  if (!editor) {
    alert("No active note editor detected. Please open workspace.");
    sharedObj.error = true;
  }
  return "";
})()}
// @beforeloop-end
// @default-begin
${await new Promise(async (r) => {
  if (sharedObj.error) {
    return;
  }

  if (topItem.isRegularItem()) {
    const annotations = [];
    const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
      i.isPDFAttachment()
    );
    attachments.forEach((i) => {
      annotations.splice(0, 0, ...i.getAnnotations());
    });
    const notes = Zotero.Items.get(topItem.getNotes());
    annotations.splice(0, 0, ...notes);
    annotations.forEach((annotation) =>
      annotation
        .getTags()
        .map((tag) => tag.tag)
        .forEach((tag) => {
          sharedObj.updateData(annotation, tag);
        })
    );
    Zotero.Items.get(topItem.getNotes()).forEach((note) =>
      note
        .getTags()
        .map((tag) => tag.tag)
        .forEach((tag) => {
          sharedObj.updateData(note, tag);
        })
    );
  } else if (topItem.isNote()) {
    topItem
      .getTags()
      .map((tag) => tag.tag)
      .forEach((tag) => {
        sharedObj.updateData(topItem, tag);
      });
  }
  return r("");
})}
// @default-end
// @afterloop-begin
${await new Promise(async (r) => {
  if (sharedObj.error) {
    return;
  }
  console.log(sharedObj);
  let res = "";
  for (const tag in sharedObj.tagsDict) {
    res += `<h1>Tag: ${tag}</h1>\n`;
    for (const note of sharedObj.tagsDict[tag].filter((item) =>
      item.isNote()
    )) {
      res += `<p><a href="${Zotero.Knowledge4Zotero.knowledge.getNoteLink(
        note
      )}">${note.getNoteTitle()}</a></p>\n<p> <blockquote>${note.getNote()} </blockquote></p>`
	  copyNoteImage(note)
    }
    res += await Zotero.Knowledge4Zotero.NoteParse.parseAnnotationHTML(
      editor._item,
      sharedObj.tagsDict[tag].filter((item) => item.isAnnotation())
    );
    res += "\n";
  }
  return r(res);
})}
// @afterloop-end
```

**Comment by sjgknight** [comment]
Caveat, this might work for a relatively small number of short notes, but due to some zotero note constraints it will probably fail for anything large.

**Comment by censalata** [comment]
Did you ever get this to work? I think this is exactly what I need.
